alf/algorithms/sac_algorithm.py

```diff
@@ -248,40 +258,27 @@ def rollout_step(self, time_step: TimeStep, state):
                                       "is not supported by SacAlgorithm")
         return self._predict(time_step, state, epsilon_greedy=1.0)
 
+    def _compute_critics(self, critic_net, observation, action, critics_state):
+        if self._use_q_network:
+            # critics shape: [B, replicas, num_actions]
+            critics, critics_state = critic_net(
+                observation, state=critics_state)
+        else:
+            # critics shape: [B, replicas]
+            critics, critics_state = critic_net((observation, action),
+                                                state=critics_state)
+        return critics, critics_state
+
     def _actor_train_step(self, exp: Experience, state: SacActorState,
                           action_distribution, action, log_pi):
 
-        if self._is_continuous:
-            critic_input = (exp.observation, action)
-
-            critics, critics_state = self._critic_networks(
-                critic_input, state=state.critics)
-
-            target_q_value = critics.min(dim=1)[0]
-            dqda = nest.pack_sequence_as(
-                action,
-                list(
-                    torch.autograd.grad(
-                        target_q_value.sum(),
-                        nest.flatten(action),
-                    )))
-
-            def actor_loss_fn(dqda, action):
-                if self._dqda_clipping:
-                    dqda = torch.clamp(dqda, -self._dqda_clipping,
-                                       self._dqda_clipping)
-                loss = 0.5 * losses.element_wise_squared_loss(
-                    (dqda + action).detach(), action)
-                loss = loss.sum(list(range(1, loss.ndim)))
-                return loss
-
-            actor_loss = nest.map_structure(actor_loss_fn, dqda, action)
-            alpha = torch.exp(self._log_alpha).detach()
-            actor_loss += alpha * log_pi
-        else:
-            critics, critics_state = self._critic_networks(
-                exp.observation, state=state.critics)
+        critics, critics_state = self._compute_critics(
+            self._critic_networks, exp.observation, action, state.critics)
+        alpha = torch.exp(self._log_alpha).detach()
 
+        if self._use_q_network:
+            # Make the entire Categorical dist shape close to the target Q value
```

Comment:
For this case, p(a) have a close form solution p(a) = exp(q(a)) / alpha, which should be much better. Because of this, for mixed case, it's better to model the q values as Q(a_d|a_c, s), where a_c is continuous action, and a_d is discrete action.

Reply:
Yeah adding a_c as a condition will theoretically avoid the non-stationarity issue. But I guess we need to do some experiment to see how this compares with the common critic solution.

Reply:
For the pure discrete case, it's almost certainly better because there is no need to train an actor network. Thus it's likely to better for the mixed case if the discrete action is a scalar.
Currently, out main use cases for discrete action is still the scalar. We can deal with multidimensional discrete action when we actually encounter it.

Reply:
The actor network as a function approximator still has to be trained across different states s, even though there is a closed form solution for every given Q(a_d|s). It's just that for every training loss, the entire distribution is considered because we could enumerate all the probability entries since it's discrete and finite. SGD is still needed to gradually adjust the actor given Q values. I think the inefficiency of having a critic network for discrete actions is mainly due to the empirical estimation of the actor loss instead of an accurate calculation of the actor loss. So the question is really how much error will be introduced by this empirical estimation vs. introduced by a conditional Q(a_d|a_c,s).

Reply:
For discrete case, we don't need actor loss because the action distribution is exp(Q(a|s) / alpha). For mixed case, the loss is only for the continuous action a_c. Once a_c is sampled, a_d is based on exp(Q(a_d|a_c, s)/alpha).

Reply:
Still, the learning of discrete actions will make the learning of continuous actions non-stationary because the dependency is one-way in Q(a_d|a_c,s). That is, the same continuous action made under the same state would generate different rewards and critic values. Do you think this is a problem?